### PR TITLE
Improve yaws_ls:trim/3 to trim at UTF8 char boundary

### PR DESCRIPTION
**Pull request by leoliu**
Before this commit:

```
世界足球d运动会开始了全世界��..>
```

After this commit:

```
世界足球d运动会开始了全世界都..>
```

### CODE REVIEW

**Comment by vinoski** [comment]
It would be good to also have a test for this.

File: test/eunit/yaws_ls_tests.erl [hunk 1]
```diff
@@ -0,0 +1,13 @@
+%% @doc `yaws_ls_tests' -- eunit tests for `yaws_ls'
+
```
Comment (vinoski) [inline test/eunit/yaws_ls_tests.erl]: I would get rid of lines 1 and 2 here and replace them with `%% -*- coding: latin-1 -*-` (or whatever setting is appropriate).

File: test/eunit/yaws_ls_tests.erl [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+-module(yaws_ls_tests).
+-include_lib("eunit/include/eunit.hrl").
+
+trim_test(Expected, Input, Len) ->
+    Enc = binary_to_list(unicode:characters_to_binary(Input)),
+    ExpectedEnc = binary_to_list(unicode:characters_to_binary(Expected)),
+    ExpectedEnc =:= yaws_ls:trim(Enc, Len).
+
+trim_test_() ->
+    [?_assert(trim_test("世界杯..&gt;", "世界杯足球赛", 11)),
```
Comment (vinoski) [inline test/eunit/yaws_ls_tests.erl]: This line makes anything older than Erlang 17 very unhappy in the Travis builds.

**Comment by leoliu** [comment]
Hi @vinoski,

I am not sure if the amended commit will fix the issue though. I have no OTP < 17 on my systems.

**Comment by vinoski** [comment]
@leoliu: it's easy to support multiple versions of Erlang on your system with [`kerl`](https://github.com/kerl/kerl).

With the latest update it looks like all versions older than R16B are still failing.

**Comment by vinoski** [comment]
Maybe @capflam can have a look? I don't know enough about the details of unicode support in various Erlang versions to know what the problem might be.

**Comment by capflam** [comment]
Yes of course, I will check

**Comment by leoliu** [comment]
@vinoski, thanks for that. I already have many versions of erlang installed but nothing beyond R16B. 
@capflam thanks.

**Comment by capflam** [comment]
Hi,

Well, it is a problem with the Eunit test.  Since release R16,  literal strings in compiled modules are stored using their unicode representation. Older releases store them using their bytewise representation. If you compile the following module with different releases, you will not have always the same result.

```erlang
-module(test).
-export([t/0]).

t() -> io:format("~p~n~p~n~p~n", [<<"世界杯足球赛"/utf8>>, <<"世界杯足球赛">>, "世界杯足球赛"]).
```

On R16B03:
```erlang
<<228,184,150,231,149,140,230,157,175,232,182,179,231,144,131,232,181,155>>
<<22,76, 111,179,3,91>>
[19990,30028,26479,36275,29699,36187]
```

On R1503:
```erlang
<<195,164,194,184,194,150,195,167,194,149,194,140,195,166,194,157,194,175,195,168,194,182,194,179,195,167,194,144,194,131,195,168,194,181,194,155>>
<<228,184,150,231,149,140,230,157, 175,232,182,179,231,144,131,232,181,155>>
 [228,184,150,231,149,140,230,157, 175,232,182,179,231,144,131,232,181,155]
```

So, you should rewrite your test. Here is a quite patch that should fix the bug on older releases:

```diff
diff --git a/test/eunit/yaws_ls_tests.erl b/test/eunit/yaws_ls_tests.erl
index 0a6249c..a9148b1 100644
--- a/test/eunit/yaws_ls_tests.erl
+++ b/test/eunit/yaws_ls_tests.erl
@@ -8,5 +8,9 @@ trim_test(Expected, Input, Len) ->
     ExpectedEnc =:= yaws_ls:trim(Enc, Len).
 
 trim_test_() ->
-    [?_assert(trim_test("世界杯..&gt;", "世界杯足球赛", 11)),
-     ?_assert(trim_test("xxxxxxxx..&gt;", "xxxxxxxxxxxx", 11))].
+    [
+     %% Trim "世界杯足球赛"
+     ?_assert(trim_test([19990,30028,26479|"..&gt;"],
+                        [19990,30028,26479,36275,29699,36187], 11)),
+     ?_assert(trim_test("xxxxxxxx..&gt;", "xxxxxxxxxxxx", 11))
+    ].
```

That's not pretty handy, but it works.

**Comment by leoliu** [comment]
hi @capflam,

Many thanks for that info. I amended the commit per your comment. Let's see if the tests will pass this time. Thanks again.

**Comment by vinoski** [comment]
I manually merged these changes to master. Thanks @leoliu !

**Comment by leoliu** [comment]
Thanks @vinoski and @capflam for helping out!

**Pull request closed without merging**